File: src/leetCode/Ex113.java
package leetCode;

import java.util.ArrayList;
import java.util.List;

public class Ex113 {
    private List<Integer> temps = new ArrayList<>();
    private List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        helper(root, sum, 0);
        return res;
    }

    private void helper(TreeNode root, int sum, int temp){
        if(root==null){
            return;
        }
        if(root.left==null&&root.right==null&&temp+root.val==sum){
            temps.add(root.val);
            res.add(new ArrayList<>(temps));
            temps.remove(temps.size()-1);
            return;
        }
        temps.add(root.val);
        helper(root.left, sum, temp+root.val);
        helper(root.right, sum, temp+root.val);
        temps.remove(temps.size()-1);
    }
}

File: src/leetCode/ex113.kt
package leetCode

import java.util.*
import kotlin.collections.HashSet

fun pathSum(root: TreeNode?, sum: Int): List<List<Int>> {
    val res = mutableListOf<List<Int>>()
    val stack = Stack<TreeNode>()
    if (root==null){
        return res
    }
    val tempPath = mutableListOf<Int>()
    stack.push(root)
    val usedSet = HashSet<TreeNode>()
    while (stack.isNotEmpty()){
        val t = stack.peek()
        if (usedSet.contains(t)){
            tempPath.removeAt(tempPath.size-1)
            stack.pop()
            continue
        }
        usedSet.add(t)
        tempPath.add(t.`val`)
        if (t.right!=null){
            stack.push(t.right)
        }
        if (t.left!=null){
            stack.push(t.left)
        }
        if (t.left==null&&t.right==null){
            //t is leave node
            if (tempPath.sum()==sum) {
                val cp = mutableListOf<Int>().apply {
                    tempPath.forEach { this.add(it) }
                }
                res.add(cp)
            }
        }
    }
    return res
}


